shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,837 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
+        Ok(self.store.contains(&Self::parse_key(key)?))
+    }
+
+    /// Update the tree with a new value.
+    ///
+    /// Returns the new root if successful. Will
+    /// return an error if the key is malformed.
+    pub fn insert_key(&mut self, key: &Key) -> Result<Hash, Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.insert(hash);
+        self.root = self.compute_root();
+        Ok(self.root())
+    }
+
+    /// Delete a key from storage and update the root
+    pub fn delete_key(&mut self, key: &Key) -> Result<(), Error> {
+        let hash = Self::parse_key(key)?;
+        _ = self.store.remove(&hash);
+        self.root = self.compute_root();
+        Ok(())
+    }
+
+    /// Compute the root of the merkle tree
+    fn compute_root(&self) -> KeccakHash {
+        let mut hashes: Vec<KeccakHash> = self.store.iter().cloned().collect();
+        while hashes.len() > 1 {
+            let mut next_hashes = vec![];
+            for pair in hashes.chunks(2) {
+                let left = pair[0].clone();
+                let right = pair.get(1).cloned().unwrap_or_default();
+                next_hashes.push(hash_pair(left, right));
+            }
+            hashes = next_hashes;
+        }
+
+        if hashes.is_empty() {
+            Default::default()
+        } else {
+            hashes.remove(0)
+        }
+    }
+
+    /// Return the root as a [`struct@Hash`] type.
+    pub fn root(&self) -> Hash {
+        self.root.clone().into()
+    }
+
+    /// Get a reference to the backing store
+    pub fn store(&self) -> &BTreeSet<KeccakHash> {
+        &self.store
+    }
+
+    /// Create a batched membership proof for the provided keys
+    pub fn get_membership_proof(
+        &self,
+        keys: &[Key],
+        mut values: Vec<PendingTransfer>,
+    ) -> Result<BridgePoolProof, Error> {
+        if values.len() != keys.len() {
+            return Err(eyre!(
+                "The number of leaves and leaf hashes must be equal."
+            )
+            .into());
+        }
+        // sort the values according to their hash values
+        values.sort_by_key(|transfer| transfer.keccak256());
+
+        // get the leaf hashes
+        let mut leaves: BTreeSet<KeccakHash> = Default::default();
+        for (key, value) in keys.iter().zip(values.iter()) {
+            let hash = Self::parse_key(key)?;
+            if hash != value.keccak256() {
+                return Err(eyre!(
+                    "Hashes of keys did not match hashes of values."
+                )
+                .into());
+            }
+            leaves.insert(hash);
+        }
+
+        let mut proof_hashes = vec![];
+        let mut flags = vec![];
+        let mut hashes: Vec<_> = self
+            .store
+            .iter()
+            .cloned()
+            .map(|hash| {
+                if leaves.contains(&hash) {
+                    Node::OnPath(hash)
+                } else {
+                    Node::Sibling(hash)
```

Comment:
Re. the `Node::Sibling` docstring, we could just call this `Node::NotOnPath` or something. Here some of the leaves we're wrapping could be not siblings of any of the `leaves` we want to make a proof for?

Reply:
Sure.